kernel_tuner/cache/convert.py

```diff
@@ -0,0 +1,250 @@
+
+
+def unversioned_convert(cache: dict, schema_path: Path) -> dict:
+    """Attempts a conversion of an unversioned cache file to version 1.0.0.
+
+    Parameters:
+        ``cache`` is a ``dict`` representing the cachefile.
+
+    Returns:
+        a ``dict`` representing the converted cache file.
+
+    Raises:
+        ``ValueError`` if given cache file is too old and no suitable
+        conversion exists.
+    """
+    cache["schema_version"] = "1.0.0"
+
+    if "objective" not in cache:
+        cache["objective"] = "time"
+
+    for key, entry in cache["cache"].items():
+        if "timestamp" not in entry:
+            cache["cache"][key]["timestamp"] = "2024-06-18 11:36:56.137831+00:00"
+        for missing_key in ["compile_time", "benchmark_time", "framework_time", "strategy_time"]:
+            if missing_key not in entry:
+                cache["cache"][key][missing_key] = 0
+
+    path = schema_path / "1.0.0/schema.json"
+
+    with open(path) as s:
+        versioned_schema = json.load(s)
+
+    missing_keys = [key for key in versioned_schema["properties"] if key not in cache]
+    if missing_keys:
+        raise ValueError(
+            f"Cache file too old, missing key{'s' if len(missing_keys) > 1 else ''} "
+            + f"{', '.join(missing_keys)}, no suitable conversion to version 1.0.0 exists."
+        )
+
+    return cache
+
+
+def convert_cache_to_t4(cache: CacheFileJSON) -> T4FileJSON:
+    """Converts a cache file to the T4 auto-tuning format.
+
+    ``cache`` is a ``CacheFileJSON`` representing the cache file to convert.
+
+    Returns a ``T4FileJSON`` representing the converted cache file.
+    """
+    t4 = T4FileJSON(results=[], schema_version="1.0.0")
+
+    for cache_line in cache["cache"].values():
+        times = T4ResultTimesJSON(
+            compilation_time=cache_line["compile_time"],
+            framework=cache_line["framework_time"],
+            search_algorithm=cache_line["strategy_time"],
+            validation=cache_line["verification_time"],
+        )
+        # In case of invalid configurations 'times' might be omitted in cache_line
+        if "times" in cache_line:
+            times["runtimes"] = cache_line["times"]
+
+        measurement = T4ResultMeasurementJSON(name=cache["objective"], value=cache_line[cache["objective"]], unit="")
+
+        # Convert Kernel Tuner's ErrorConfig instances to 'invalidity' strings supported in T4
+        invalidity = "correct"
+        objective_value = cache_line[cache["objective"]]
+        if isinstance(objective_value, str):
+            # The kernel configuration is invalid
+            if objective_value == "CompilationFailedConfig":
+                invalidity = "compile"
+            elif objective_value == "RuntimeFailedConfig":
+                invalidity = "runtime"
+            elif objective_value == "InvalidConfig":
+                invalidity = "constraints"
+
+        result = T4ResultLineJSON(
+            timestamp=cache_line["timestamp"],
```

Comment:
This line requires the input cachefile to have timestamps, is that something we want to enforce or are we okay with something like `cache_line.get("timestamp", None)`?